libraries/chain/finality/finalizer.cpp

```diff
@@ -105,6 +118,41 @@ vote_message_ptr finalizer::maybe_vote(const bls_public_key& pub_key,
 }
 
 // ----------------------------------------------------------------------------------------
+inline bool has_voted_strong(const std::vector<finalizer_authority>& active_finalizers, const valid_quorum_certificate& qc, const bls_public_key& key) {
+   assert(qc.is_strong());
+   auto it = std::find_if(active_finalizers.begin(),
+                          active_finalizers.end(),
+                          [&](const auto& finalizer) { return finalizer.public_key == key; });
+
+   if (it != active_finalizers.end()) {
+      auto index = std::distance(active_finalizers.begin(), it);
+      assert(qc.strong_votes);
+      return qc.strong_votes->test(index);
+   }
+   return false;
+}
+
+void my_finalizers_t::maybe_update_fsi(const block_state_ptr& bsp, const valid_quorum_certificate& received_qc) {
+   if (finalizers.empty())
+      return;
+
+   assert(bsp->active_finalizer_policy);
+
+   // see comment on possible optimization in maybe_vote
+   std::lock_guard g(mtx);
```

Comment:
Can this be moved down right before line 152 `save_finalizer_safety_info()`?

Reply:
The mutex guards modifications of `finalizer.fsi`. It could be moved down before the first call to `maybe_update_fsi` but didn't think the extra complexity was worth it. Would need to then see if the mutex was locked on each iteration of the loop as you could not unlock it until the file is written.